docs/engineering-principles.md

```diff
@@ -0,0 +1,48 @@
+# Engineering Principles
+Engineering principles define the core tenets and beliefs that the engineering team and all contributors are expected to keep front of mind. These tenets are meant to reflect the values of the organization as well as the philosophy upon which the software was founded. Principles typically come in the form of punchy one line statements, but a great deal of depth can be written on each one to further inform of the intent of the statement.
+
+## Core Engineering Values
+These one word statements are used to not only capture the qualities we strive for in our software, but also prioritizes those values. Keep in mind that being the last value in this list does not imply it is of no importance, it is still more important than all of the other qualities not mentioned here. 
+1. Security
+2. Stability
+3. Safety
+4. Privacy
+5. Maintainability
+6. Usability
+
+
+## Principles
+### The Principle of Least Authority
+
+Provide application code the absolute minimal amount of authority required to perform its essential functions. Wherever appropriate, break down large methods with broad authority into smaller functions with minimal authority. Minimizing what a method *can* do reduces the scope of what could possibly go wrong. When something does go wrong, small isolated methods, written with the least amount of authority required, are easier to troubleshoot.
+
+### The Principle of Core Purity
```

Comment:
Was this meant to say "Code purity"?

Reply:
No, there's a "principle" of software development called "functional sandwich" which describes keeping the "core" or middle of the application layer pure and functional, and moving imperative code to the outer layer. I dislike the phrase functional sandwich and was trying to be crafty with the words but still  keeping the sentiment. Could you suggest an alternative? We could make it "Code Purity" as it definitely fits and is something i'd like for us to strive for.

Reply:
I'm having trouble applying this principle to our codebase. Most of us probably conceptualize the controllers as "core" to the codebase for example, and their primary purpose is to be _the place_ where we handle mutable state.

I'm not sure what the "outer edge" of the UI would be either :thinking:

Reply:
Maybe this advice is too specific to be at this level? Or maybe there is a different way to describe where stateful things should happen, instead of "outer edges of the application layer"?

Reply:
Changed this to code purity and removed sentiment around layers/core in favor of just "majority of code should be pure" as a goal

Reply:
Hmm. This now seems even more out of place now.

Can we remove this principle for now? It seems like more of a stylistic preference than a principle. It's a preference that I share, but this doesn't seem like the appropriate place for it. Maybe it's better suited to being in a guide on how to make your code easier to read, or a guide on how to design APIs to be more easily tested.

Reply:
I agree that it's appropriate for "best practices". I was thinking of "best practices" as falling under guides but we could definitely split them up if that would be helpful.

Reply:
https://github.com/MetaMask/contributor-docs/pull/6/commits/74988a28113866d59c1cebcafaffbcbc28d6d691 I removed this principle here.